Connection closed without response, with an Authentication script with errors

Steps to reproduce the issue:
1. Run ZAP (version 2.4.2);
2. Create an "Authentication" script with a call to an undefined function;
3. Setup the authentication of a context to use the created authentication script and enable "Forced User";
4. Proxy a request;
5. Note that the user-agent doesn't get a response, but it should even if the authentication was not successful;
6. Check the log file, there should be an exception:
with Nashorn (Java 8):
```
1138140 [ZAP-ProxyThread-10] ERROR org.zaproxy.zap.ZAP$UncaughtExceptionLogger  - Exception in thread "ZAP-ProxyThread-10"
<eval>:18 ReferenceError: "undefinedFunction" is not defined
	at jdk.nashorn.internal.runtime.ECMAErrors.error(ECMAErrors.java:57)
	at jdk.nashorn.internal.runtime.ECMAErrors.referenceError(ECMAErrors.java:319)
	at jdk.nashorn.internal.runtime.ECMAErrors.referenceError(ECMAErrors.java:291)
	at jdk.nashorn.internal.objects.Global.__noSuchProperty__(Global.java:1432)
	at jdk.nashorn.internal.scripts.Script$Recompilation$42$1253AAA$\^eval\_.authenticate(<eval>:18)
	at org.zaproxy.zap.authentication.ScriptBasedAuthenticationMethodType$AuthenticationScript$$NashornJavaAdapter.authenticate(Unknown Source)
	at org.zaproxy.zap.authentication.ScriptBasedAuthenticationMethodType$ScriptBasedAuthenticationMethod.authenticate(Unknown Source)
	at org.zaproxy.zap.users.User.authenticate(Unknown Source)
	at org.zaproxy.zap.users.User.processMessageToMatchUser(Unknown Source)
	at org.parosproxy.paros.network.HttpSender.sendAuthenticated(Unknown Source)
	at org.parosproxy.paros.network.HttpSender.sendAndReceive(Unknown Source)
	at org.parosproxy.paros.network.HttpSender.sendAndReceive(Unknown Source)
	at org.parosproxy.paros.core.proxy.ProxyThread.processHttp(Unknown Source)
	at org.parosproxy.paros.core.proxy.ProxyThread.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
```
or, with Rhino (Java 7):
```
697137 [ZAP-ProxyThread-15] ERROR org.zaproxy.zap.ZAP$UncaughtExceptionLogger  - Exception in thread "ZAP-ProxyThread-15"
java.lang.reflect.UndeclaredThrowableException
	at com.sun.proxy.$Proxy5.authenticate(Unknown Source)
	at org.zaproxy.zap.authentication.ScriptBasedAuthenticationMethodType$ScriptBasedAuthenticationMethod.authenticate(Unknown Source)
	at org.zaproxy.zap.users.User.authenticate(Unknown Source)
	at org.zaproxy.zap.users.User.processMessageToMatchUser(Unknown Source)
	at org.parosproxy.paros.network.HttpSender.sendAuthenticated(Unknown Source)
	at org.parosproxy.paros.network.HttpSender.sendAndReceive(Unknown Source)
	at org.parosproxy.paros.network.HttpSender.sendAndReceive(Unknown Source)
	at org.parosproxy.paros.core.proxy.ProxyThread.processHttp(Unknown Source)
	at org.parosproxy.paros.core.proxy.ProxyThread.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.security.PrivilegedActionException: javax.script.ScriptException: sun.org.mozilla.javascript.internal.EcmaError: ReferenceError: "undefinedFunction" is not defined. (<Unknown source>#18) in <Unknown source> at line number 18
	at java.security.AccessController.doPrivileged(Native Method)
	at com.sun.script.util.InterfaceImplementor$InterfaceImplementorInvocationHandler.invoke(InterfaceImplementor.java:67)
	... 10 more
Caused by: javax.script.ScriptException: sun.org.mozilla.javascript.internal.EcmaError: ReferenceError: "undefinedFunction" is not defined. (<Unknown source>#18) in <Unknown source> at line number 18
	at com.sun.script.javascript.RhinoScriptEngine.invoke(RhinoScriptEngine.java:300)
	at com.sun.script.javascript.RhinoScriptEngine.invokeFunction(RhinoScriptEngine.java:258)
	at com.sun.script.util.InterfaceImplementor$InterfaceImplementorInvocationHandler$1.run(InterfaceImplementor.java:70)
	... 12 more
Caused by: sun.org.mozilla.javascript.internal.EcmaError: ReferenceError: "undefinedFunction" is not defined. (<Unknown source>#18)
	at sun.org.mozilla.javascript.internal.ScriptRuntime.constructError(ScriptRuntime.java:3770)
	at sun.org.mozilla.javascript.internal.ScriptRuntime.constructError(ScriptRuntime.java:3748)
	at sun.org.mozilla.javascript.internal.ScriptRuntime.notFoundError(ScriptRuntime.java:3833)
	at sun.org.mozilla.javascript.internal.ScriptRuntime.getNameFunctionAndThis(ScriptRuntime.java:2218)
	at sun.org.mozilla.javascript.internal.Interpreter.interpretLoop(Interpreter.java:1510)
	at sun.org.mozilla.javascript.internal.Interpreter.interpret(Interpreter.java:849)
	at sun.org.mozilla.javascript.internal.InterpretedFunction.call(InterpretedFunction.java:162)
	at sun.org.mozilla.javascript.internal.ContextFactory.doTopCall(ContextFactory.java:430)
	at com.sun.script.javascript.RhinoScriptEngine$1.superDoTopCall(RhinoScriptEngine.java:116)
	at com.sun.script.javascript.RhinoScriptEngine$1.doTopCall(RhinoScriptEngine.java:109)
	at sun.org.mozilla.javascript.internal.ScriptRuntime.doTopCall(ScriptRuntime.java:3160)
	at sun.org.mozilla.javascript.internal.InterpretedFunction.call(InterpretedFunction.java:160)
	at com.sun.script.javascript.RhinoScriptEngine.invoke(RhinoScriptEngine.java:294)
	... 14 more
```

ZAP Version:
Versions >= 2.3.0

--- a/src/org/zaproxy/zap/authentication/ScriptBasedAuthenticationMethodType.java
+++ b/src/org/zaproxy/zap/authentication/ScriptBasedAuthenticationMethodType.java
@@ -206,7 +206,9 @@ public class ScriptBasedAuthenticationMethodType extends AuthenticationMethodTyp
 				script = getScriptsExtension().getInterface(this.script, AuthenticationScript.class);
 				msg = script.authenticate(new AuthenticationHelper(getHttpSender(), sessionManagementMethod,
 						user), this.paramValues, cred);
-			} catch (ScriptException | IOException e) {
+			} catch (Exception e) {
+				// Catch Exception instead of ScriptException and IOException because script engine implementations
+				// might throw other exceptions on script errors (e.g. jdk.nashorn.internal.runtime.ECMAException)
 				log.error("An error occurred while trying to authenticate using the Authentication Script: "
 						+ this.script.getName(), e);
 				getScriptsExtension().setError(this.script, e);

